addons/mail/controllers/mail.py

```diff
@@ -97,7 +97,17 @@ def _redirect_to_record(cls, model, res_id, access_token=None, **kwargs):
         else:
             record_action = record_sudo.get_access_action()
             if record_action['type'] == 'ir.actions.act_url' and record_action.get('target_type') != 'public':
-                return cls._redirect_to_messaging()
+                url_params = {
+                    'model': model,
+                    'id': res_id,
+                    'active_id': res_id,
+                    'action': record_action.get('id'),
+                }
+                view_id = record_sudo.get_formview_id()
```

Comment:
Why not rebuilding a mail/view-based link ? That way if a portal user arrives while not being logged, he is redirected to a portal view instead of crashing (as he cannot access backend). Would make some sort of "try again" based on original link...